README.md

```diff
@@ -1,2 +1,81 @@
+  Ready            True    Sat, 10 Apr 2021 02:23:21 +0530   Fri, 09 Apr 2021 15:45:45 +0530   KubeletReady                 kubelet is posting ready status
+
+  ```
+- Following are the NodeConditions by default supported by K8s cluster.
+  
+  
+  | ConditionType      | Condition Status   |Effect        | Key      |
+  | ------------------ | ------------------ | ------------ | -------- |
+  |Ready               |True                | -            | |
+  |                    |False               | NoExecute    | node.kubernetes.io/not-ready           |
+  |                    |Unknown             | NoExecute    | node.kubernetes.io/unreachable         |
+  |OutOfDisk           |True                | NoSchedule   | node.kubernetes.io/out-of-disk         |
+  |                    |False               | -            | |
+  |                    |Unknown             | -            | |
+  |MemoryPressure      |True                | NoSchedule   | node.kubernetes.io/memory-pressure     |
+  |                    |False               | -            | |
+  |                    |Unknown             | -            | |
+  |DiskPressure        |True                | NoSchedule   | node.kubernetes.io/disk-pressure       |
+  |                    |False               | -            | |
+  |                    |Unknown             | -            | |
+  |NetworkUnavailable  |True                | NoSchedule   | node.kubernetes.io/network-unavailable |
+  |                    |False               | -            | |
+  |                    |Unknown             | -            | |
+  |PIDPressure         |True                | NoSchedule   | node.kubernetes.io/pid-pressure        |
+  |                    |False               | -            | |
+  |                    |Unknown             | -            | |
+
+
+- `Node Problem Detector`
+  - By default, k8s support limited set of NodeConditions. 
+  - We can use [node-problem-detector](https://github.com/kubernetes/node-problem-detector) which runs as a DaemonSet and collects different node problems and reports them in form of NodeConditions.
+  
+- Based on the NodeCondition's we can apply taint effects such as
+  - `NoSchedule`: Does not allow new pods to schedule onto the node unless they tolerate the taint. Does not interrupt already running pods.
+  - `PreferNoSchedule`: Scheduler tries not to schedule new pods onto the node.
+  - `NoExecute`: Evicts any already-running pods that do not tolerate the taint. 
+  
+
+## Node Taint Controller(NTC)
+- NTC control loop will continuously look for `node conditions` of all nodes and will apply taint based on the `condition` type. 
+
+## Node Auto Healer Operator(NAHO)
+- Will reconcile all Nodes, look for applied `taints` on each node and apply taint effect based on the taint type on that node.
+- Next, it will evict all the pods from that Node.
```

Comment:
Drain node one by one in the same node group. We can use go channel to lock the drain process so we can guarantee only one node is drained at one time.

Reply:
Yes, we can use go's channel to achieve synchronization in the draining process. And just to keep record of which node is being drained we can refer to `HealedNode`'s `currentState`, which would be helpful to regain channels previous state just in case the controller restarts. 
We can also provide support of parallel recovering if enabled using CR [NodeAutoHealer](https://github.com/stakater/jarvis/blob/add-readme/docs/node-auto-healer-operator.md#nodeautohealer)'s `parallelHealing` field and we can initialize size of buffered channel with `maxAllowedNodesToDrain` count.